apisix/plugins/hmac-auth.lua

```diff
@@ -182,13 +201,25 @@ local function generate_signature(ctx, secret_key, params)
         canonical_query_string = core.table.concat(query_tab, "&")
     end
 
-    local req_body = core.request.get_body()
-    req_body = req_body or ""
+    local canonical_headers = ""
+
+    core.log.info("all headers: ", core.json.encode(core.request.headers(), true))
+
+    for _, h in pairs(params.signed_headers) do
+        canonical_headers = canonical_headers .. (core.request.header(ctx, h) or "")
```

Comment:
`canonical_headers` can be an array object, then we can use `table.insert` and `table.concat` way, which is better performance

Reply:
not quite understand here, this is concatenating a string to generate a signature

Reply:
I helped you fix it. you can take a look when you have time